designs/dotnetcore-debugging.md

```diff
@@ -0,0 +1,196 @@
+
+And let's break it down to the checklist
+
+* Put the `vsdbg` on the Docker container;
+* Add _"wait for the debugger"_ flag in runner program and implement corresponding functionality;
+* Implement corresponding container entry point override in SAM to support invoking the runner with this flag **on**.
+* Provide instructions for the customer of how to configure VS Code and VS to attach.
+
+### What will be changed
+
+Runner program for .NET core [2.0](https://github.com/lambci/docker-lambda/blob/master/dotnetcore2.0/run/MockBootstraps/Program.cs) and [2.1](https://github.com/lambci/docker-lambda/blob/master/dotnetcore2.1/run/MockBootstraps/Program.cs) in another [repo](https://github.com/lambci/docker-lambda) to support waiting for the debugger to attach. Already have [PR](https://github.com/lambci/docker-lambda/pull/127) (guys and @mhart - reviewers wanted :)). I think we should provide new `sam local` flag  `--container-name` which will assign desired name to the container with the Lambda runtime to simplify `docker exec` experience from VS Code later.
+
+### Success Criteria
+
+1. Provide handy script to get the `vsdbg` via downloading it to the .NET Lambda runtime container, installing it and getting through the mounting (or manual instructions to achieve the above results). After that user would be able to provide it to SAM via `--debugger-path`;
+2. Provide ready-to-go `launch.json` configuration for attaching to the Lambda function via VSCode;
+3. Customer should be able to easily debug .NET Core 2.0 and 2.1 apps on VS Code (minimal requirement).
+
+### Out-of-Scope
+
+Visual Studio 2017 (can be a target, but it has some other very different Docker container debugging support - needs additional investigation). Help in this area is highly appreciated.
+
+### User Experience Walkthrough
+
+##### 1. Getting the debugger locally
+
+At first, user should be able to get and install `vsdbg` on their machine easily. For this we should provide the script for them to invoke, which will effectively run the .NET Lambda runtime Docker container with mounted via `bind` path to get the debugger on the **host** machine. In this container following: `curl -sSL https://aka.ms/getvsdbgsh | bash /dev/stdin -v latest -l /vsdbg`  should be run to get and install required debugger (specifically built for this container). Later SAM will mount this same folder with the debugger to the target running container. 
+
+Instead of script we can certainly provide manual instructions of how to do the above stuff step-by-step as it is now done for golang debugger. _Feedback appreciated_.
+
+Customer should go through this step **only once** to set up the debugger.
```

Comment:
is my understanding correct, that we just need to be able to download the debugger to the host machine. I dont think there is any setup required, just a mount into the docker container right?

Yeah in that case, just downloading the ```vsdbg``` debugger onto one path., only once is enough.

Reply:
Yes, I've meant the path of getting and installing it via Docker. User would have to go along it only once. Just wanted to point that out.

Reply:
And, actually, setup is required, because we are just downaloding `getvsdbg.sh` which needs to be executed in order to have `vsdbg` built for specific system. That is why we are doing that through the Docker container of the actual  runtime, on which debugging will occur.